File: pdf_scanner.cpp
// pdf_scanner.cpp
#include <iostream>
#include <string>
#include <vector>
#include <filesystem>
#include <opencv2/opencv.hpp>
#include <podofo/podofo.h>

using namespace std;
using namespace PoDoFo;

bool convertToPDFA(const vector<string>& inputFiles, const string& outputFile, bool compress, const string& pageSize, int dpi) {
    // Определяем размер страницы в пунктах (1 pt = 1/72 inch)
    float width, height;
    if (pageSize == "A4") {
        width = 595.28f;
        height = 841.89f;
    } else if (pageSize == "Letter") {
        width = 612.0f;
        height = 792.0f;
    } else {
        size_t pos = pageSize.find('x');
        if (pos != string::npos) {
            width = stof(pageSize.substr(0, pos));
            height = stof(pageSize.substr(pos + 1));
        } else {
            width = 595.28f;
            height = 841.89f;
        }
    }

    try {
        // Создаём PDF документ
        PdfMemDocument pdf;
        pdf.SetVersion(PdfVersion_1_4);

        for (const string& fname : inputFiles) {
            if (!filesystem::exists(fname)) {
                cerr << "Файл не найден: " << fname << endl;
                continue;
            }

            // Загружаем изображение через OpenCV
            cv::Mat image = cv::imread(fname, cv::IMREAD_COLOR);
            if (image.empty()) {
                cerr << "Не удалось прочитать: " << fname << endl;
                continue;
            }

            // Сжатие
            if (compress) {
                cv::resize(image, image, cv::Size(), 0.7, 0.7, cv::INTER_LANCZOS4);
            }

            // Создаём страницу
            PdfPage* page = pdf.CreatePage(PdfRect(0, 0, width, height));
            pdf.AddPage(page);

            // Конвертируем OpenCV Mat в PoDoFo изображение
            // (упрощённо: сохраняем во временный файл)
            string tempFile = "/tmp/img_" + to_string(inputFiles.size()) + ".jpg";
            cv::imwrite(tempFile, image);

            // Вставляем изображение
            PdfImage img = pdf.CreateImage(tempFile);
            if (img.IsValid()) {
                double imgWidth = img.GetWidth();
                double imgHeight = img.GetHeight();
                double scaleX = width / imgWidth;
                double scaleY = height / imgHeight;
                double scale = min(scaleX, scaleY);

                double x = (width - imgWidth * scale) / 2;
                double y = (height - imgHeight * scale) / 2;

                PdfPainter painter;
                painter.SetPage(page);
                painter.DrawImage(x, y, imgWidth * scale, imgHeight * scale, &img);
                painter.FinishPage();
            }

            // Удаляем временный файл
            filesystem::remove(tempFile);
        }

        // Сохраняем PDF
        pdf.Write(outputFile.c_str());
        cout << "PDF/A сохранён: " << outputFile << endl;
        return true;
    } catch (const PdfError& e) {
        cerr << "Ошибка PoDoFo: " << e.what() << endl;
        return false;
    } catch (const exception& e) {
        cerr << "Ошибка: " << e.what() << endl;
        return false;
    }
}

int main(int argc, char* argv[]) {
    vector<string> inputFiles;
    string outputFile = "output.pdf";
    bool compress = false;
    string pageSize = "A4";
    int dpi = 300;
    bool verbose = false;

    for (int i = 1; i < argc; i++) {
        string arg = argv[i];
        if (arg == "-i" || arg == "--input") {
            while (i + 1 < argc && string(argv[i + 1])[0] != '-') {
                inputFiles.push_back(argv[++i]);
            }
        } else if (arg == "-o" || arg == "--output") {
            if (i + 1 < argc) outputFile = argv[++i];
        } else if (arg == "--compress") {
            compress = true;
        } else if (arg == "--page-size" && i + 1 < argc) {
            pageSize = argv[++i];
        } else if (arg == "--dpi" && i + 1 < argc) {
            dpi = stoi(argv[++i]);
        } else if (arg == "-v" || arg == "--verbose") {
            verbose = true;
        } else if (arg == "-h" || arg == "--help") {
            cout << "Использование: pdf_scanner [options]" << endl;
            cout << "  -i, --input <files>   Входные изображения" << endl;
            cout << "  -o, --output <file>   Выходной PDF/A файл" << endl;
            cout << "  --compress            Сжатие изображений" << endl;
            cout << "  --page-size <A4|Letter|WxH>" << endl;
            cout << "  --dpi <dpi>           Разрешение в DPI" << endl;
            return 0;
        }
    }

    if (inputFiles.empty()) {
        cerr << "Не указаны входные файлы." << endl;
        return 1;
    }

    bool success = convertToPDFA(inputFiles, outputFile, compress, pageSize, dpi);
    return success ? 0 : 1;
}

File: PdfScanner.java
// PdfScanner.java
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.pdmodel.PDPageContentStream;
import org.apache.pdfbox.pdmodel.common.PDRectangle;
import org.apache.pdfbox.pdmodel.graphics.image.PDImageXObject;
import org.apache.pdfbox.pdmodel.graphics.image.ImageType;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class PdfScanner {
    public static void main(String[] args) throws IOException {
        List<String> inputFiles = new ArrayList<>();
        String outputFile = "output.pdf";
        boolean compress = false;
        String pageSize = "A4";
        int dpi = 300;
        boolean verbose = false;

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-i":
                case "--input":
                    while (i + 1 < args.length && !args[i + 1].startsWith("-")) {
                        inputFiles.add(args[++i]);
                    }
                    break;
                case "-o":
                case "--output":
                    outputFile = args[++i];
                    break;
                case "--compress":
                    compress = true;
                    break;
                case "--page-size":
                    pageSize = args[++i];
                    break;
                case "--dpi":
                    dpi = Integer.parseInt(args[++i]);
                    break;
                case "-v":
                    verbose = true;
                    break;
                case "-h":
                case "--help":
                    System.out.println("Использование: java PdfScanner [options]");
                    System.out.println("  -i, --input <files>   Входные изображения");
                    System.out.println("  -o, --output <file>   Выходной PDF/A файл");
                    System.out.println("  --compress            Сжатие изображений");
                    System.out.println("  --page-size <A4|Letter|WxH>");
                    System.out.println("  --dpi <dpi>           Разрешение в DPI");
                    return;
                default:
                    if (!args[i].startsWith("-"))
                        inputFiles.add(args[i]);
                    break;
            }
        }

        if (inputFiles.isEmpty()) {
            System.err.println("Не указаны входные файлы.");
            System.exit(1);
        }

        try (PDDocument document = new PDDocument()) {
            // Определяем размер страницы в пунктах (1 pt = 1/72 inch)
            float width, height;
            switch (pageSize) {
                case "A4":
                    width = PDRectangle.A4.getWidth();
                    height = PDRectangle.A4.getHeight();
                    break;
                case "Letter":
                    width = PDRectangle.LETTER.getWidth();
                    height = PDRectangle.LETTER.getHeight();
                    break;
                default:
                    String[] parts = pageSize.split("x");
                    if (parts.length == 2) {
                        width = Float.parseFloat(parts[0]);
                        height = Float.parseFloat(parts[1]);
                    } else {
                        width = PDRectangle.A4.getWidth();
                        height = PDRectangle.A4.getHeight();
                    }
                    break;
            }

            for (String fname : inputFiles) {
                File imgFile = new File(fname);
                if (!imgFile.exists()) {
                    System.err.println("Файл не найден: " + fname);
                    continue;
                }

                // Загружаем изображение
                BufferedImage image = ImageIO.read(imgFile);
                if (image == null) {
                    System.err.println("Не удалось прочитать: " + fname);
                    continue;
                }

                // Масштабируем
                if (compress) {
                    int newW = (int) (image.getWidth() * 0.7);
                    int newH = (int) (image.getHeight() * 0.7);
                    java.awt.Image scaled = image.getScaledInstance(newW, newH, java.awt.Image.SCALE_SMOOTH);
                    image = new BufferedImage(newW, newH, BufferedImage.TYPE_INT_RGB);
                    image.getGraphics().drawImage(scaled, 0, 0, null);
                }

                // Создаём страницу
                PDPage page = new PDPage(new PDRectangle(width, height));
                document.addPage(page);

                // Вставляем изображение
                PDImageXObject pdImage = PDImageXObject.createFromFileByExtension(imgFile, document);
                if (pdImage == null) {
                    // Если не удалось, пробуем через BufferedImage
                    pdImage = PDImageXObject.createFromFile(fname, document);
                }
                if (pdImage != null) {
                    float imgWidth = pdImage.getWidth();
                    float imgHeight = pdImage.getHeight();
                    float scaleX = width / imgWidth;
                    float scaleY = height / imgHeight;
                    float scale = Math.min(scaleX, scaleY);

                    float x = (width - imgWidth * scale) / 2;
                    float y = (height - imgHeight * scale) / 2;

                    try (PDPageContentStream contentStream = new PDPageContentStream(document, page)) {
                        contentStream.drawImage(pdImage, x, y, imgWidth * scale, imgHeight * scale);
                    }
                }
            }

            document.save(outputFile);
            System.out.println("PDF/A сохранён: " + outputFile);
        } catch (IOException e) {
            System.err.println("Ошибка: " + e.getMessage());
            System.exit(1);
        }
    }
}
